### PR TITLE
Refactor pass pipeline for better precompute globals

### LINKED ISSUE #2153: Flaky const expression optimizations
Consider this snippet:
```ts
let w = 42 >> 1;
w = (w & 0xFC) | (w >> 1);

export const x = w;
```

The only exported value is `x` and it's a constant, so the compiled output should likely be
```wat
 (global $module/x i32 (i32.const 30))
 (export "x" (global $module/x))
```

Instead, the compiler produces the following (with `-O3`):
```wat
 (global $module/w (mut i32) (i32.const 0))
 (global $module/x (mut i32) (i32.const 0))
 (export "x" (global $module/x))
 (start $~start)
 (func $~start
  i32.const 21
  global.set $module/w
  i32.const 30
  global.set $module/w
  i32.const 30
  global.set $module/x
 )
```

This output shows several issues:
- `x` is exported as a mutable global;
- `w` exists as an internal global despite being used only for setting the `x` value once (fixed by `-O3z`);
- two consecutive `global.set` calls are used on `w` (also fixed by `-O3z`).

---

What's more interesting is that simplifying the first line leads to even more complex output.

```ts
let w = 21;
w = (w & 0xFC) | (w >> 1);

export const x = w;
```

```wat
 (global $module/w (mut i32) (i32.const 21))
 (global $module/x (mut i32) (i32.const 0))
 (export "x" (global $module/x))
 (start $~start)
 (func $~start
  (local $0 i32)
  global.get $module/w
  local.tee $0
  i32.const 252
  i32.and
  local.get $0
  i32.const 1
  i32.shr_s
  i32.or
  global.set $module/w
  global.get $module/w
  global.set $module/x
 )
```

### PR DESCRIPTION
**Pull request by MaxGraey**
Try to fix #2153

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Comment by dcodeIO** [comment]
Do you think this is worth it given the regressions?

**Comment by MaxGraey** [comment]
Well, recent commit minimize side effects

File: tests/compiler/do.optimized.wat [hunk 1]
```diff
@@ -996,14 +996,6 @@
...
   (local $1 i32)
   (local $2 i32)
-  (local $3 i32)
-  (local $4 i32)
-  (local $5 i32)
-  (local $6 i32)
-  (local $7 i32)
-  (local $8 i32)
-  (local $9 i32)
-  (local $10 i32)
```
Comment (MaxGraey) [inline tests/compiler/do.optimized.wat:1006]: That's really good stuff

File: tests/compiler/for.optimized.wat [hunk 2]
```diff
@@ -996,26 +996,19 @@
   (local $0 i32)
   (local $1 i32)
   (local $2 i32)
-  (local $3 i32)
-  (local $4 i32)
-  (local $5 i32)
-  (local $6 i32)
-  (local $7 i32)
-  (local $8 i32)
-  (local $9 i32)
```
Comment (MaxGraey) [inline tests/compiler/for.optimized.wat:1005]: again

File: tests/compiler/possibly-null.optimized.wat [hunk 3]
```diff
@@ -50,35 +50,6 @@
...
-  block $__inlined_func$possibly-null/testFalseElse
-   local.get $0
-   i32.eqz
-   br_if $__inlined_func$possibly-null/testFalseElse
-  end
-  global.get $~lib/memory/__stack_pointer
-  i32.const 4
-  i32.add
-  global.set $~lib/memory/__stack_pointer
- )
```
Comment (MaxGraey) [inline tests/compiler/possibly-null.optimized.wat:81]: also very nice

File: tests/compiler/std/array.optimized.wat [hunk 4]
```diff
@@ -14382,70 +14399,17 @@
...
-  (local $58 i32)
-  (local $59 i32)
-  (local $60 i32)
-  (local $61 i32)
-  (local $62 i32)
-  (local $63 i32)
-  (local $64 i32)
-  (local $65 i32)
-  (local $66 i32)
-  (local $67 i32)
```
Comment (MaxGraey) [inline tests/compiler/std/array.optimized.wat:14448]: Look at this!

**Review by dcodeIO — APPROVED** [review]
Heh, I'm convinced :)

**Pull request merged**